src/meepgeom.cpp

```diff
@@ -2707,99 +2799,181 @@ void material_grids_addgradient_point(double *v, std::complex<double> fields_a,
+            cyl_scale = (gv.dim == meep::Dcyl) ? 2*p.r() : 1; // the pi is already factored in near2far.cpp
+            material_grids_addgradient_point(v+ng*f_i, vec_to_vector3(p), scalegrad*prod.real()*cyl_scale, geps);
+          /* more complicated case requires interpolation/restriction */
+          }else{
+            /* we need to restrict the adjoint fields to
+            the two nodes of interest, and interpolate
+            the forward fields to the same two nodes. Then
+            we perform our inner product at these nodes
+            */            
+            std::complex<double> fwd_avg, fwd1, fwd2, prod;
+            ptrdiff_t fwd1_idx, fwd2_idx;
+
+            //identify the first corner of the forward fields
+            meep::ivec fwd_p = ip + gv.iyee_shift(forward_c) - gv.iyee_shift(adjoint_c);
+
+            //identify the other three corners
+            meep::ivec unit_a  = unit_ivec(gv.dim,component_direction(adjoint_c));
+            meep::ivec unit_f  = unit_ivec(gv.dim,component_direction(forward_c));
+            meep::ivec fwd_pa  = (fwd_p + unit_a*2);
+            meep::ivec fwd_pf  = (fwd_p - unit_f*2);
+            meep::ivec fwd_paf = (fwd_p + unit_a*2 - unit_f*2);
+            
+            //identify the two eps points
+            meep::ivec ieps1 = (fwd_p + fwd_pf) / 2;
+            meep::ivec ieps2 = (fwd_pa + fwd_paf) / 2;
+
+            //operate on the first eps node
+            fwd1_idx = get_idx_from_ivec(gv.dim, start_ivec, the_stride, fwd_p);
+            fwd1 = 0.5 * GET_FIELDS(fields_f,ci_forward,f_i,fwd1_idx);
+            fwd2_idx = get_idx_from_ivec(gv.dim, start_ivec, the_stride, fwd_pf);
+            fwd2 = 0.5 * GET_FIELDS(fields_f,ci_forward,f_i,fwd2_idx);
+            fwd_avg = fwd1 + fwd2;
+            meep::vec eps1 = gv[ieps1];
+            cyl_scale = (gv.dim == meep::Dcyl) ? eps1.r() : 1;
+            prod = 0.5*adj*get_material_gradient(eps1, adjoint_c, forward_c, fwd_avg, frequencies[f_i], geps, 1e-6);
+            material_grids_addgradient_point(v+ng*f_i, vec_to_vector3(eps1), scalegrad*prod.real()*cyl_scale, geps);
+            
+            //operate on the second eps node
+            fwd1_idx = get_idx_from_ivec(gv.dim, start_ivec, the_stride, fwd_pa);
+            fwd1 = 0.5 * GET_FIELDS(fields_f,ci_forward,f_i,fwd1_idx);
+            fwd2_idx = get_idx_from_ivec(gv.dim, start_ivec, the_stride, fwd_paf);
+            fwd2 = 0.5 * GET_FIELDS(fields_f,ci_forward,f_i,fwd2_idx);
+            fwd_avg = fwd1 + fwd2;
+            meep::vec eps2 = gv[ieps2];
+            cyl_scale = (gv.dim == meep::Dcyl) ? eps2.r() : 1;
+            prod = 0.5*adj*get_material_gradient(eps2, adjoint_c, forward_c, fwd_avg, frequencies[f_i], geps, 1e-6);
+            material_grids_addgradient_point(v+ng*f_i, vec_to_vector3(eps2), scalegrad*prod.real()*cyl_scale, geps);
```

Comment:
There is redundant code here that can be separated into a function and called twice with different arguments `(fwd_p, fwd_pf, eps1)` and `(fwd_pa, fwd_paf, eps2)`.

Reply:
I thought so too... but everything worth refactoring is already refactored in `get_material_gradient` and `material_grids_addgradient_point`.

Pulling it out further just made it more unreadable (and requires a lot more parameters to pass than just those 3).

`prod` could technically be placed within `material_grids_addgradient_point`. `cyl_scale` is repeated, but again, refactoring actually makes it harder to read.